### PR TITLE
avm2: Re-implement `parseFloat`

### PR DESCRIPTION
**Pull request by relrelb**
The AVM2 `parseFloat` implementation used to call into the AVM1 one,
which was not 100% accurate. De-couple it as a new implementation,
based on `MathUtils::convertStringToDouble` in `avmplus`:
https://github.com/adobe/avmplus/blob/master/core/MathUtils.cpp#L1229-L1439

### CODE REVIEW

**Comment by adrian17** [comment]
Can you please add some tests for edge cases and bug-compatibility (if you know any such cases)?

File: core/src/avm2/globals/toplevel.rs [hunk 1]
```diff
@@ -141,30 +141,225 @@ pub fn parse_int<'gc>(
...
+
+    let is_negative = parse_sign(&mut s);
+    let after_sign = s;
+
+    // Count digits before decimal point.
+    s = s.trim_start_matches(|c| (c as u8).is_ascii_digit());
+    let mut total_digits = after_sign.len() - s.len();
+
+    // Stop parsing at null character.
+    if s.starts_with(b'\0') {
```
Comment (kmeisthax) [inline core/src/avm2/globals/toplevel.rs:193]: Is there a reason why we can't truncate the string right at the start of the function instead of checking at specific points throughout?

**Review by kmeisthax — COMMENTED** [review]
I'm going to second the request for test cases, it will help if this code ever needs to be refactored.

File: core/src/avm2/globals/toplevel.rs (hunk 1 above)
Comment (relrelb) [reply core/src/avm2/globals/toplevel.rs:193]: That what `avmplus` did. But after thinking about it, I came into a conclusion that these stops are only relevant for the `strict` mode check below (which isn't used yet - should be used by `Number(...)`). So I updated this PR by replacing the null character stops with a `strict` mode check that permits a null character. Once `Number(...)` will use `string_to_f64` in a future PR, `Number("10" + String.fromCharCode(0))` will be considered valid.
